src/lib/schedule.c

```diff
@@ -25,459 +25,97 @@
-			delta = deadline - current;
-
-			/* get highest priority */
-			if (task->priority < next_priority) {
-				next_priority = task->priority;
-				next_delta = delta;
-				next_task = task;
-			} else if (task->priority == next_priority) {
-				/* get earliest deadline */
-				if (delta < next_delta) {
-					next_delta = delta;
-					next_task = task;
-				}
-			}
-		} else {
-			/* missed scheduling - will be rescheduled */
-			trace_schedule("edf_get_next(), "
-				   "missed scheduling - will be rescheduled");
-
-			/* have we already tried to rescheule ? */
-			if (!reschedule) {
-				reschedule++;
-				trace_schedule("edf_get_next(), "
-					       "didn't try to reschedule yet");
-				edf_reschedule(task, current);
-			} else {
-				/* reschedule failed */
-				list_item_del(&task->list);
-				task->state = TASK_STATE_CANCEL;
-			}
-		}
+	task->type = type;
+	//TODO: Generic priority
+	task->priority = priority;
+	//task->priority = task_priority[type][priority];
+
+	task->state = SOF_TASK_INIT_STATE;
+	task->func = func;
+	task->data = data;
+
+	switch (task->type) {
+	case SOF_EDF_TASK:
+		schedule_task_init(task, core);
+		break;
+	case SOF_LOW_LATENCY_TASK:
+		ll_task_init(task, xflags);
+		break;
 	}
-
-	return next_task;
 }
 
-/* work set in the future when next task can be scheduled */
-static uint64_t sch_work(void *data, uint64_t delay)
+void schedule_sof_task(struct sof_task *task, uint64_t start,
+	uint64_t deadline, uint32_t flags)
 {
-	tracev_schedule("sch_work()");
-	schedule();
-	return 0;
-}
-
-/*
- * EDF Scheduler - Earliest Deadline First Scheduler.
- *
- * Schedule task with the earliest deadline from task list.
- * Can run in IRQ context.
- */
-static struct task *schedule_edf(void)
-{
-	struct schedule_data *sch = *arch_schedule_get();
-	struct task *task;
-	struct task *future_task = NULL;
-	uint64_t current;
-	uint32_t flags;
-
-	tracev_schedule("schedule_edf()");
-
-	interrupt_clear(PLATFORM_SCHEDULE_IRQ);
-
-	while (!list_is_empty(&sch->list)) {
-		spin_lock_irq(&sch->lock, flags);
-
-		/* get the current time */
-		current = platform_timer_get(platform_timer);
-
-		/* get next task to be scheduled */
-		task = edf_get_next(current, NULL);
-		spin_unlock_irq(&sch->lock, flags);
-
-		/* any tasks ? */
-		if (!task)
-			return NULL;
-
-		/* can task be started now ? */
-		if (task->start <= current) {
-			/* yes, run current task */
-			task->start = current;
-
-			/* init task for running */
-			spin_lock_irq(&sch->lock, flags);
-			task->state = TASK_STATE_PENDING;
-			list_item_del(&task->list);
-			spin_unlock_irq(&sch->lock, flags);
-
-			/* now run task at correct run level */
-			if (run_task(task) < 0) {
-				trace_error(TRACE_CLASS_SCHEDULE,
-					    "schedule_edf() error");
-				break;
-			}
-		} else {
-			/* no, then schedule wake up */
-			future_task = task;
-			break;
-		}
+	switch (task->type) {
+	case SOF_EDF_TASK:
+		if (flags & SOF_SCHEDULE_IDLE)
+			schedule_edf_task_idle(task, deadline);
```

Comment:
Flag should be handled internally by edf_schedule.

Reply:
Done

Reply:
done